test/integration/oauth_serviceaccount_client_events_test.go

```diff
@@ -0,0 +1,437 @@
+		RedirectUrl:  redirectURL,
+		Scope:        scope.Join([]string{"user:info", "role:edit:" + projectName}),
+		SendClientSecretInParams: true,
+	}
+
+	doOAuthFlow(t, ts.clusterAdminClientConfig, oauthClientConfig, ts.authCodes, ts.authErrors, expectBadRequest, []string{
+		"GET /oauth/authorize",
+		"received challenge",
+		"GET /oauth/authorize",
+		"redirect to /oauth/authorize/approve",
+		"form",
+		"POST /oauth/authorize/approve",
+		"redirect to /oauth/authorize",
+		"redirect to /oauthcallback",
+		"code",
+	})
+
+	ts.clusterAdminOAuthClient.OAuthClientAuthorizations().Delete(adminUser+":"+oauthClientConfig.ClientId, nil)
+}
+
+func doOAuthFlow(
+	t *testing.T,
+	clusterAdminClientConfig *restclient.Config,
+	oauthClientConfig *osincli.ClientConfig,
+	authorizationCodes chan string,
+	authorizationErrors chan string,
+	expectBadRequest bool,
+	expectOperations []string,
+) {
+	drain(authorizationCodes)
+	drain(authorizationErrors)
+
+	oauthRuntimeClient, err := osincli.NewClient(oauthClientConfig)
+	if err != nil {
+		t.Fatalf("unexpected error: %v", err)
+	}
+	clientTransport, err := restclient.TransportFor(clusterAdminClientConfig)
+	testTransport := &basicAuthTransport{rt: clientTransport}
+	oauthRuntimeClient.Transport = testTransport
+
+	authorizeRequest := oauthRuntimeClient.NewAuthorizeRequest(osincli.CODE)
+	req, err := http.NewRequest("GET", authorizeRequest.GetAuthorizeUrlWithParams("").String(), nil)
+	if err != nil {
+		t.Fatalf("unexpected error: %v", err)
+	}
+
+	// Set up the HTTP redirect handler
+	operations := []string{}
+	jar, _ := cookiejar.New(nil)
+	directHTTPClient := &http.Client{
+		Transport: testTransport,
+		CheckRedirect: func(redirectReq *http.Request, via []*http.Request) error {
+			t.Logf("302 Location: %s", redirectReq.URL.String())
+			req = redirectReq
+			operations = append(operations, "redirect to "+redirectReq.URL.Path)
+			return nil
+		},
+		Jar: jar,
+	}
+
+	for {
+		t.Logf("%s %s", req.Method, req.URL.String())
+		operations = append(operations, req.Method+" "+req.URL.Path)
+
+		// Always set the csrf header
+		req.Header.Set("X-CSRF-Token", "1")
+		resp, err := directHTTPClient.Do(req)
+		if err != nil {
+			t.Errorf("Error %v\n%#v\n%#v", operations, jar, err)
+			return
+		}
+		defer resp.Body.Close()
+
+		// Save the current URL for reference
+		currentURL := req.URL
+
+		if resp.StatusCode == 401 {
+			// Set up a username and password once we're challenged
+			testTransport.username = adminUser
+			testTransport.password = "any-pass"
+			operations = append(operations, "received challenge")
+			continue
+		}
+
+		if expectBadRequest && resp.StatusCode == 400 {
+			responseDump, _ := httputil.DumpResponse(resp, true)
+			t.Logf("Bad Request: %s", string(responseDump))
+			return
+		}
+
+		if resp.StatusCode != 200 {
+			responseDump, _ := httputil.DumpResponse(resp, true)
+			t.Errorf("Unexpected response %s", string(responseDump))
```

Comment:
Consider including to the error message that we expected to have `200` status code.